autopilot/contractor.go

```diff
@@ -546,12 +549,24 @@ func (c *contractor) runContractChecks(ctx context.Context, w Worker, contracts
 			continue
 		}
 
-		// fetch recent price table and attach it to host.
-		host.PriceTable, err = c.priceTable(ctx, w, host.Host)
-		if err != nil {
-			c.logger.Errorf("could not fetch price table for host %v: %v", host.PublicKey, err)
-			toStopUsing[fcid] = "could not fetch price table"
-			continue
+		// if the host doesn't have a valid pricetable, update it
+		if time.Now().After(host.PriceTable.Expiry) {
+			host.PriceTable, err = c.priceTable(ctx, w, host.Host)
+			if err != nil {
+				c.logger.Errorf("could not fetch price table for host %v: %v", host.PublicKey, err)
+
+				// NOTE: if we can not fetch the price table, we give the host
+				// the benefit of the doubt if it's in the current set and we
+				// have some leeway
+				_, ok := inCurrentSet[fcid]
```

Comment:
I think we shouldn't give any host the benefit of the doubt before performing the `isUsableHost` checks.
Assuming the host has a valid price table and fails any of the host checks, we'd drop it from the set. Not doing so just because the host is missing the price table doesn't seem very consistent.

So what I'd do here is something like this:

```go
// if the host doesn't have a valid pricetable, update it
hasValidPT := true
if time.Now().After(host.PriceTable.Expiry) {
	pt, err := c.priceTable(ctx, w, host.Host)
	if err != nil {
		c.logger.Errorf("could not fetch price table for host %v: %v", host.PublicKey, err)
                hasValidPT = false
        } else {
                host.PriceTable = pt
        }
}

// Perform `isUsableHost` check. In the worst case on the outdated pricetable.

if contract.Revision == nil || !hasPriceTable {
   // handle accordingly with leeway
}

Reply:
I updated it to be more in line with what you have in mind. It's more or less the same but I ended up splitting up the checks.